### PR TITLE
Hardware wallet signer fixes #139

### LINKED ISSUE #139: Implement hardware wallet transaction signing strategy
**Feature description**

Sign transactions with Skycoin hardware wallet

**Is your feature request related to a problem? Please describe.**
Signing transactions with hardware wallet is not yet supported.

**Describe the solution you'd like**
Implement transaction signing using [Skycoin hardware wallet go library](https://github.com/skycoin/hardware=wallet-go) .

**Describe alternatives you've considered**
[Skycoin hardware wallet daemon](https://github.com/skycoin/hardware=wallet-daemon) service is another way to go , but over-complicates an initial integration . it also adds non trivial complexity and security risks .

**Additional context**
Requires #138 .

**Possible implementation**

- [x] Implement `TxnSignStrategy` for SkyWallet devices .

### PR DESCRIPTION
**Pull request by stdevAlDen**
Fixes #139

**Changes:**
- Remove `vendor/github.com/SkycoinProject/skycoin` git submodule, use it from vendors.
- Run [skywallet](https://github.com/fibercrypto/skywallet-mcu) in travis to run integration tests against it.
- Fix bug in `GetInputs`, previous cached value.
- Add [skywallet-go](github.com/fibercrypto/skywallet-go) dependency
- Implement signer iterator interface.
- GetSignerUID and GetSignerDescription could return error.
- Implement `EnumerateSignServices` for local/remote wallet.
- Implement get wallet matching the first address
- Add `TestTransactionSignInputFromDevice` for the hardware wallet signer.
- Implement signer for hardware wallet.
- Add `QSigner` and `SignerModel` models
- Add gui to select signer in simple mode for sending

**Does this change need to mentioned in CHANGELOG.md?**
yes

**Requires changes in protobuf specs?**
no

**Requires testing**
yes

### CODE REVIEW

File: .travis.yml
Comment (olemis) [inline .travis.yml]: could you please implement this as `make` target or CI script ?

File: .travis.yml
Comment (olemis) [inline .travis.yml]: move to `before_script`

**Review by olemis — CHANGES_REQUESTED** [review]
Move `src/hardware` => `src/contrib/skywallet`

File: Makefile
Comment (olemis) [inline Makefile]: Rename `test-hw` => `test-skyhw`

File: ci-scripts/integration-test.sh
Comment (olemis) [inline ci-scripts/integration-test.sh]: What's this file for ?

File: src/coin/skycoin/models/coin.go
Comment (olemis) [inline src/coin/skycoin/models/coin.go]: Use logger instance . Ask @stdevMac if in doubt .

File: src/coin/skycoin/models/cipher.go
Comment (olemis) [inline src/coin/skycoin/models/cipher.go]: Not sure about object cloning methods . Are those still needed?

File: src/coin/skycoin/models/signer_iterator.go
Comment (olemis) [inline src/coin/skycoin/models/signer_iterator.go]: Duplicated functionality ? Same as `DefaultTxnSignerIterator` ?

File: src/coin/skycoin/models/sky_test.go
Comment (olemis) [inline src/coin/skycoin/models/sky_test.go:1]: Functions in this file have to remain separate from he rest , since they've been copied verbatim from Skycoin code base .

File: src/coin/skycoin/models/skyapi_mock_test.go
Comment (olemis) [inline src/coin/skycoin/models/skyapi_mock_test.go]: Functions in this file have to remain apart of the rest due to their specific purpose of mocking real classes .

File: src/coin/skycoin/models/testsutil.go
Comment (olemis) [inline src/coin/skycoin/models/testsutil.go]: Remove commented package imports ?

File: src/coin/skycoin/models/testsutil.go
Comment (olemis) [inline src/coin/skycoin/models/testsutil.go]: Rename `testsutil.go` => `testutils.go`

File: src/coin/skycoin/models/testsutil.go
Comment (olemis) [inline src/coin/skycoin/models/testsutil.go]: This is not specific to Skycoin altcoin plugin . Please move onto `util` package.

File: src/coin/skycoin/models/wallet.go
Comment (olemis) [inline src/coin/skycoin/models/wallet.go]: is this tested ?

File: src/coin/skycoin/models/wallet.go
Comment (olemis) [inline src/coin/skycoin/models/wallet.go]: What's the final word on this ?

File: src/core/coin.go
Comment (olemis) [inline src/core/coin.go]: Not sure this still needed . Remove ?

File: src/core/coin.go
Comment (olemis) [inline src/core/coin.go]: Cloning capabilities are not requirements of transactions.

File: src/core/coin.go
Comment (olemis) [inline src/core/coin.go]: Cloning capabilities are not requirements of transactions.

File: src/core/coin.go
Comment (olemis) [inline src/core/coin.go]: Cloning capabilities are not requirements of transactions.

**Review by olemis — CHANGES_REQUESTED** [review]

File: src/core/wallet.go
Comment (olemis) [inline src/core/wallet.go]: Rename `GetWallet()` => `LookupWallet()`

File: src/errors/error.go
Comment (olemis) [inline src/errors/error.go]: Start error message with capital letters.

File: src/hardware/sky-wallet.go
Comment (olemis) [inline src/hardware/sky-wallet.go]: Shouldn't this be `signer:skywallet` ?

File: src/hardware/sky-wallet.go
Comment (olemis) [inline src/hardware/sky-wallet.go]: Do not use logrus directly , but loggers . Ask @stdevMac if in doubt .

File: src/hardware/sky-wallet.go
Comment (olemis) [inline src/hardware/sky-wallet.go]: No translation for log messages.

File: src/hardware/sky-wallet.go
Comment (olemis) [inline src/hardware/sky-wallet.go]: What about these lines commented out ?

File: src/hardware/sky-wallet.go
Comment (olemis) [inline src/hardware/sky-wallet.go]: Expand argument type scope . All instances of `SkycoinTxn` implement `EncodeSkycoinTransaction()`

**Review by olemis — CHANGES_REQUESTED** [review]

**Comment by olemis** [comment]
@stdevAlDen does code need to be updated after merging fibercrypto/skywallet-go#164 ?

File: .gitignore
Comment (stdevMac) [inline .gitignore:5]: @stdevAlDen I think we need this in the gitignore, cause some of the deps in therecipe/qt are needed and static, example `widgets`. Any special reason to remove it?

**Review by stdevMac — CHANGES_REQUESTED** [review]

File: .gitignore
Comment (stdevAlDen) [reply .gitignore:5]: > @stdevAlDen I think we need this in the gitignore, cause some of the deps in therecipe/qt are needed and static, example `widgets`. Any special reason to remove it?

@stdevMac when therecipe qt binding is inside the vendors the project does not build (at least for me), no other thing...

Comment (olemis) [reply .gitignore:5]: but adding it in `.gitignore` hides the fact that there is something wrong with the working copy .

File: src/coin/skycoin/models/wallet.go
Comment (olemis) [inline src/coin/skycoin/models/wallet.go:828]: @stdevAlDen why this change ?

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go:12]: `fce` => `fcwerrors`

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go:131]: Duplicated . Reuse the logic in Skycoin altcoin plugin .

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go:192]: Duplicated . Reuse Skycoin altcoin plugin code .

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go:206]: What happens on error ?

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go]: Duplicate . Reuse Skycoin altcoin plugin code.

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go:218]: Duplicate . Reuse Skycoin altcoin plugin code

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go:246]: Why sign failure if not signing transaction yet ?

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go:255]: Use different error conditions in each case to identify failure path.

Comment (olemis) [reply src/contrib/skywallet/sky-wallet.go:255]: Let me take a second try on it .

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go:445]: No error , but generic placeholder ...

File: src/contrib/skywallet/sky-wallet.go
Comment (olemis) [inline src/contrib/skywallet/sky-wallet.go:1]: Package name should match folder name . `skywallet` ?

File: src/core/wallet.go
Comment (olemis) [inline src/core/wallet.go:27]: Need be renamed .

File: src/models/walletsManager.go
Comment (olemis) [inline src/models/walletsManager.go:860]: condition should be `err == nil` , right ?

File: src/models/walletsModel.go
Comment (olemis) [inline src/models/walletsModel.go:220]: What about this `FIXME` tag ? Shall this remain ?

File: test/integration/util/hardware-wallet.go
Comment (olemis) [inline test/integration/util/hardware-wallet.go:1]: Shouldn't this be under `skywallet` contrib package ?

**Review by olemis — CHANGES_REQUESTED** [review]

**Pull request merged**
